### ISSUE TITLE
[ENH] `polars` support - as `sktime` datatype

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#5423]
Issue for initial design discussion and planning.

Mid-term, I think it would be good to support `polars`.
`scikit-learn` is currently adding `polars` support, and even considering to loosen the dependency on `pandas`.
We should probably consider the same.

In terms of technical steps we would need to take, these are:

* add `polars` based mtypes for `Series`, `Panel`, and `Hierarchical` data. These can probably be best based off the `dask` mtypes.
* add `polars` or data backend agnostic implementations for the most crucial `sktime` estimators, these are probably the compositions like pipelines, reduction
* in estimator packages upstream that already support `polars`, enable the mtype and check that both inputs work
* possibly, add more scenarios to testing with `polars` input and expected output
* bonus: upgrade `skpro` to `polars` support, which will allow probabilistic estimators to be looped through `sktime` as well, and direct interaction with `sklearn` `polars` based functionality

**Comment by fkiraly** [comment]
Is there someone who can explain to me the precise requirement for `pyarrow` when using `polars`?

It seems it is required from some conversions to/between `pandas`, but not for vanilla use - this strikes me as a little strange. Also, is that correct?

**Comment by fkiraly** [comment]
Prototype framework support works in `skpro`: https://github.com/sktime/skpro/pull/130

This can be replicated for `sktime` and will require `scikit-base 0.6.1` or higher.

**Comment by fkiraly** [comment]
Added requirement discovered through the prototype https://github.com/sktime/skpro/pull/130: the `deep_equals` utility needs to be upgraded so it can check `polars` data containeres for equality, in testing.

**Comment by pranavvp16** [comment]
Working on this as mentioned in #6381. Will start by working on conversions and checkers by adding required functions in the polars adapter.

**Comment by Garve** [comment]
How will (multi) indices be handled? By passing some keywords like `level_cols` and `date_col`?

**Comment by Me-Priyank** [comment]
@fkiraly i had a look at where this stands on current `main`, since the mtype PRs (#6485, #6552, #6697) have all landed.

The mtypes themselves work. What I did not expect is that transformers are the only estimator type that still rejects polars:

```python
check_is_scitype(pl_frame, scitype=["Series", "Panel", "Hierarchical"])  # True, mtype pl.DataFrame
Detrender().fit_transform(pl_frame)                                      # TypeError
```

Forecasters, detectors, clusterers and classifiers all gate on scitype, so they picked up the polars mtypes for free when those landed. `BaseTransformer` is the only one gating on a hardcoded mtype list, `ALLOWED_INPUT_MTYPES` at `transformations/base/_base.py:192`. That list was last touched in #2219 in March 2022, about two and a half years before the polars mtypes existed, so I do not think polars was deliberately left out of it.

I added the three polars mtypes to that list locally and ran every transformer with no unmet soft deps. 62 of 99 work straight away and return polars. Panels go through the vectorization path fine, so this does not look blocked on #9680:

```
Detrender().fit_transform(polars_panel)  ->  polars.DataFrame, shape (24, 3)
```

Of the 37 that fail, 13 are missing conversion edges, mostly `pl.DataFrame -> np.ndarray`. The other 20 all come from `convert_Series_to_Panel` in `datatypes/_series_as_panel/_convert.py`, which by design assumes one of the three pandas/numpy Series mtypes and skips checks for speed (#5036). With polars in the allowlist it reaches the fallthrough and leaves `obj_mtype` unbound. So that function needs either polars handling or an explicit error there, as part of this step rather than separately.


Before sending anything for the transformer side: would you rather the polars mtypes simply get added to `ALLOWED_INPUT_MTYPES`, or have `BaseTransformer` move to scitype gating like the other base classes, so the list stops needing per mtype maintenance? The second is less to maintain but widens what transformers accept, so it seemed worth asking first.

**Comment by Me-Priyank** [comment]
@fkiraly  Breakdown of the transformer failures, on `main`, 101 transformers with no unmet soft dependencies, univariate polars Series input. 64 pass, 37 fail, and one of those is my test harness rather than polars (`ElbowClassSum` needs `y` in `fit`, fails the same way on pandas). So 36 real failures in four modes.

**20, `UnboundLocalError` at `datatypes/_series_as_panel/_convert.py:76`**
Transformers whose inner scitype is Panel or Hierarchical take the "case 2: higher scitype supported" branch at `transformations/base/_base.py:1221`, which calls `convert_Series_to_Panel(..., return_to_mtype=True)`. That function handles `pd.Series`, `pd.DataFrame` and `np.ndarray` and deliberately skips checks for speed (#5036), so polars reaches the fallthrough with `obj_mtype` never assigned. Needs a polars branch, and probably an explicit error on the fallthrough either way.
Examples: `ColumnConcatenator`, `DWTTransformer`, `PCATransformer`, `PaddingTransformer`, `ClustererAsTransformer`.

**13, `NotImplementedError` at `datatypes/_convert.py:168`**
Missing conversion edges. 11 are `pl.DataFrame -> np.ndarray` on the way in, for transformers with `X_inner_mtype="np.ndarray"`, and 2 are `np.ndarray -> pl.DataFrame` on the way back out.
Examples: `LogTransformer`, `PAA`, `SAX`, `ScaledLogitTransformer`, `TabularToSeriesAdaptor`.

**3, `TypeError` at `datatypes/_check.py:609`, output side**
There is a second hardcoded list besides `ALLOWED_INPUT_MTYPES`: `ALLOWED_OUT_MTYPES = ["pd.DataFrame", "pd.Series", "np.ndarray"]` at `transformations/base/_base.py:1361`. Composites whose inner transformer has already returned polars fail this check.
`ColumnEnsembleTransformer`, `MultiplexTransformer`, `TransformByLevel`.

**One scope note that matters for the end-to-end test.** Transformers only fail when the transformer is the outermost object. Inside a forecasting pipeline the forecaster converts polars to pandas at its own boundary first. I instrumented `Detrender._fit` in `Detrender() * NaiveForecaster()` and it receives a `pandas.DataFrame`, and the pipeline returns polars correctly today:

```
Detrender() * NaiveForecaster()   on polars  ->  OK, returns polars
Detrender()                       on polars  ->  TypeError
Detrender() * LogTransformer()    on polars  ->  TypeError
```

So a full forecasting pipeline test passes on current `main`. Still worth adding as a regression guard, but it will not reproduce this.

Happy to take this. Plan would be one PR covering the two allowlists, the `pl.DataFrame <-> np.ndarray` edges and the `convert_Series_to_Panel` branch, since "transformers accept polars" is only true once all four are done, plus the pipeline regression test. Say if you would rather have it split.

**Comment by fkiraly** [comment]
@Me-Priyank, great analysis, thanks for setting agents onto this!

Over all, I think:
* the manual list should not be curated in the base module, but obtained from the `datatypes` module registry, e.g,. `ALL_SERIES_MTYPES` etc. I vaguely remember that some mtypes were ambiguous so they needed to be removed - but nothing prevents us from putting the list of all mytpes there and removing something like `AMBIGUOUS_MTYPES`. Either way, now that there is a tag system for mtypes, that may be also an option.
* for the missing edges, I was unaware. I thought the graph should be connected, so `convert` would succeed. It would be appreciated if you could open an issue specifically about the failing polars to numpy conversion with a minimal example not involving transformers but only the `datatypes` module.
* `convert_Series_to_Panel` I would extend, and similarly series/hierarchical/panel individual conversions, for `polars` based datatypes. This is another issue to solve.
* ultimately, we should add an end-to-end test for each of the individual failure modes (there are two?) and develop against them.

**Comment by Me-Priyank** [comment]
> * ultimately, we should add an end-to-end test for each of the individual failure modes (there are two?) and develop against them.

@fkiraly there are three, not two, and one of them is part of your first bullet rather than separate.

* polars to numpy conversion edges, your second bullet
* `convert_Series_to_Panel` fallthrough, your third bullet
* the output side of the base module: there is a second hardcoded list, `ALLOWED_OUT_MTYPES = ["pd.DataFrame", "pd.Series", "np.ndarray"]` at `transformations/base/_base.py:1361`. Three composites hit it because their inner transformer has already returned polars. It needs the same registry treatment as the input list.

(The fourth row in my earlier table was my harness, `ElbowClassSum` needs `y` in `fit` and fails the same on pandas.)

On deriving from the registry, `ALL_TIME_SERIES_MTYPES` already does exactly what you describe, it is `set(all_tsm) - AMBIGUOUS_MTYPES` at `datatypes/__init__.py:82`, and `AMBIGUOUS_MTYPES` is `['numpyflat', 'alignment_loc', 'pd-long', 'pd-wide']`, which matches the three commented out entries. The tag system agrees too, `internal_only=True` is set on exactly `numpyflat`, `pd-long` and `pd-wide`.

One catch before we switch to it. That takes the list from 8 mtypes to 19, and seven of the eleven newly admitted ones have no conversion edges at all:

| mtype | conversion edges |
|---|---|
| `pl.DataFrame` / `polars_panel` / `polars_hierarchical` | 4 / 12 / 2 |
| `dask_series` / `dask_hierarchical` | 0 / 0 |
| all four `gluonts_*` | 0 |
| `xr.DataArray` | 0 |

So "not ambiguous" is not sufficient on its own, it would let dask, gluonts and xarray straight in and they would fail on the first conversion. Either the list needs a second condition, reachability in the conversion graph or a new tag, or those edges land first. Do you have a preference? A tag such as `capability:convertible` would fit the existing system, but it duplicates information the conversion graph already has.

Opening the two issues now.

**Comment by Me-Priyank** [comment]
@fkiraly  Both open: #11271 for the missing conversion edges, #11272 for the cross-scitype converters.

Starting on those two. Holding the base module list change until you weigh in on the gating question above.

**Comment by Me-Priyank** [comment]
Correction to my table above: the zero conversion edges for `dask_*`, `gluonts_*` and `xr.DataArray` were an artefact of my environment, where those packages were not installed, so their converters were never registered. With them installed, all three are connected through `pd.DataFrame`, and the Series graph on `main` has 60 of 81 edges rather than 20. So on the gating question, `ALL_TIME_SERIES_MTYPES` is less risky than my table suggested. The polars to numpy gap itself was real.

That gap is fixed in #11294. With polars locally added to the transformer allowlist, all 13 transformers that failed with `no conversion defined` now get past conversion. 12 pass, and `TransformIf` moves on to the `ALLOWED_OUT_MTYPES` output check, so what remains is the `convert_Series_to_Panel` fallthrough (#11272) and the two base module lists.

@fkiraly PTAL
